Jump to python
==

표준 라이브러리
--

* 파이썬 표준 라이브러리
>  - 유용한 파이썬 프로그램을 모아 놓은 것
>  - '라이브러리' = '도서관' 으로 원하는 정보를 찾아보는 곳
>  - 어떤 일을 할 때, 어떤 라이브러리를 사용해야 하는지 알아야 함

>  - 파이썬 표준 라이브러리는 파이썬 설치 시, 자동으로 컴퓨터에 설치됨
>  - sys, re 모듈은 파이썬의 중요한 표준 라이브러리

* datetime.date
>  - 연, 월, 일로 날짜 표현 시 사용하는 함수

>>  - A군과 B양이 2021년 12월 14일 부터 만나기 시작. 2023년 4월 5일은 둘이 사귄 지 며칠째 되는 날일까?
>>  - 사귀기 시작한 2021년 12월 14일은 무슨 요일이었을까?

In [2]:
import datetime
day1 = datetime.date(2021, 12, 14)
day2 = datetime.date(2023, 4, 5)

# 연, 월, 일로 datetime.date 객체 만들기

In [6]:
diff = day2 - day1
diff.days

477

>>  - day2에서 day1을 빼면, datatime 모듈의 timedelta 객체가 리턴

In [7]:
type(diff)

datetime.timedelta

>>  - 요일은 datatime.date 객체의 weekday 함수 사용
>>  - 0은 월요일, 1은 화요일... 6은 일요일

In [9]:
day = datetime.date(2021, 12, 14)
day.weekday()

1

>>  - isoweekday 함수로 1은 월요일, 2는 화요일... 7은 일요일 리턴

In [10]:
day.isoweekday()

2

* time 모듈의 함수 알아보기

* time.time()
>  - UTC(universal time coordinated, 협정 세계 표준시)를 사용해, 현재 시간을 실수 형태로 리턴

In [12]:
import time

time.time()

1687776484.8979087

* time.localtime
>  - time.localtime은 time.time()이 리턴한 실숫값을 사용해, 연, 월, 일, 시, 분, 초...의 형태로 바꾸어 줌

In [13]:
time.localtime(time.time())

time.struct_time(tm_year=2023, tm_mon=6, tm_mday=26, tm_hour=19, tm_min=48, tm_sec=59, tm_wday=0, tm_yday=177, tm_isdst=0)

* time.asctime
>  - time.asctime은 time.localtime가 리턴된 튜플 형태의 값을 인수로 받아, 날짜와 시간을 알아보기 쉬운 형태로 리턴하는 함수

In [14]:
time.asctime(time.localtime(time.time()))

'Mon Jun 26 19:51:18 2023'

* time.ctime
>  - time.asctime(time.localtime(time.time()))은 time.ctime()으로 표시 가능
>  - asctime과 다른 점 : 항상 현재 시간만 리턴

In [15]:
time.ctime()

'Mon Jun 26 19:52:21 2023'

* time.strftime
>  - strftime 함수는 시간에 관계된 것을 세밀하게 표현하는 여러 포맷 코드 제공
>  - time.strftime('출력할 형식 포맷 코드', time.localtime(time.time()))

In [ ]:
time.strftime('출력할 형식 포맷 코드', time.localtime(time.time()))

* 시간에 관계된 것을 표현하는 포맷 코드
>  - %a : 요일의 줄임말 ex. Mon
>  - %A : 요일 ex. Monday
>  - %b : 달의 줄임말 ex. Jan
>  - %B : 달 ex. January
>  - %c : 날짜와 시간을 출력 ex. Thu May 25 10:13:52 2023
>  - %d : 일(day) ex. [01,31]
>  - %H : 시간(hour) 24시간 출력 형태 ex. [00,23]
>  - %I : 시간(hour) 12시간 출력 형태 ex. [01,12]
>  - %j : 1년 중 누적 날짜 ex. [001,366]
>  - %m : 달 ex. [01,12]
>  - %M : 분 ex. [01, 59]
>  - %p : AM or PM ex. AM
>  - %S : 초 ex. [00,59]
>  - %U : 1년 중 누적 주(일요일 시작) ex. [00,53]
>  - %w : 숫자로 된 요일 ex. [0(일), 6(토)]
>  - %W : 1년 중 누적 주(월요일 시작) ex.[00,53]
>  - %x : 현재 설정된 지역에 기반한 날짜 출력 ex. 05/25/23
>  - %X : 현재 설정된 지역에 기반한 날짜 출력 ex. 17:22:21
>  - %Y : 연도 출력 ex. 2023
>  - %Z : 시간대 출력 ex. 대한민국 표준시
>  - %% : 문자 % ex. %
>  - %y : 세기 부분을 제외한 연도 출력 ex. 01

In [18]:
import time
time.strftime('%x', time.localtime(time.time()))

'06/26/23'

In [19]:
time.strftime('%c', time.localtime(time.time()))

'Mon Jun 26 21:20:04 2023'

* time.sleep
>  - 주로 루프 안에서 많이 사용
>  - 일정한 시간 간격을 두고 루프 실행 가능
>  - time.sleep 함수의 인수는 실수 형태

In [21]:
# sleep1.py

import time

for i in range(10):
    print(i)
    time.sleep(1)

# 1초 간격으로 0부터 9까지의 숫자 출력

0
1
2
3
4
5
6
7
8
9


In [22]:
import time

for i in range(10):
    print(i)
    time.sleep(0.5)

0
1
2
3
4
5
6
7
8
9


* 인수 없이 time 함수 사용
>  - time.localtime, time.asctime, time.strftime 함수는 입력 인수 없이 사용 가능
>  - 입력 인수 없이 사용 시, 현재 시각 기준으로 함수 수행

In [23]:
time.localtime()

time.struct_time(tm_year=2023, tm_mon=6, tm_mday=26, tm_hour=21, tm_min=25, tm_sec=11, tm_wday=0, tm_yday=177, tm_isdst=0)

In [24]:
time.asctime()

'Mon Jun 26 21:25:20 2023'

In [25]:
time.strftime('%c')

'Mon Jun 26 21:25:33 2023'

* math.gcd
>  - 최대 공약수(gcd, greatest common divisor) 쉽게 구할 수 있음
>  - 공약수? 두 수 이상의 여러 수의 공통된 약수. 공약수 중 가장 큰 수가 최대 공약수
>  - ex. 30과 15의 약수는 1, 3, 5, 15. 최대 공약수는 15

>>  - 어린이집에서 사탕 60개, 초콜릿 100개, 젤리 80개 준비. 똑같이 나누어 봉지에 담을 때, 최대 몇 봉지까지 만들 수 있을까?

In [27]:
import math

math.gcd(60, 100, 80)

# python 3.9 미만 버전에서는 2개까지 허용되어 에러가 뜸

'''

(knh1) knh@gnode5:~$ python -V
Python 3.8.17

'''

# python 3.9 버전 부터는 math.gcd에 여러 개의 인수 입력 가능

TypeError: gcd expected 2 arguments, got 3

In [3]:
import math

math.gcd(60, 100, 80)

'''

python 3.9 version인 가상환경 새로 만들자

conda create -n py3.9 python=3.9

(base) knh@gnode5:~$ conda activate py3.9                                                                                                                                                                                                                                                                  
(py3.9) knh@gnode5:~$ 

'''

20

In [4]:
60/20, 100/20, 80/20

(3.0, 5.0, 4.0)

* math.lcm
>  - 최소 공배수(lcm, least common multiple) 구할 때 사용
>  - 최소 공배수란 두 수의 공통 배수 중 가장 작은 수
>  - ex. 3과 5의 최소 공배수는 15

>>  - 버스 정류장에 시내버스는 15분마다 도착, 마을 버스는 25분마다 도착. 오후 1시에 두 버스 동시 도착. 두 버스 동시 도착할 다음 시각은?

In [5]:
import math

math.lcm(15, 25)

# 75분 후인 오후 2시 15분

75

* random
>  - 난수(규칙이 없는 임의의 수) 발생시키는 모듈

In [8]:
import random

random.random()

# 0.0에서 1.0 사이의 실수 중 난수 값 리턴

0.48702385653011115

In [9]:
random.randint(1, 10)

# 1에서 10 사이의 정수 중 난수 값 리턴

1

In [10]:
random.randint(1, 55)

45

In [11]:
# random_pop.py

import random

def random_pop(data):
    number = random.randint(0, len(data)-1)
    return data.pop(number)

if __name__=="__main__":
    data = [1, 2, 3, 4, 5]
    while data:
        print(random_pop(data))

# random_pop 함수는 리스트의 요소 중 무작위로 하나 선택해 꺼내고, 그 값을 리턴
# 꺼낸 요소는 pop 메서드에 의해 사라짐

2
5
4
1
3


In [13]:
import random

def random_pop(data):
    number = random.choice(data)
    data.remove(number)
    return number

if __name__=="__main__":
    data = [1, 2, 3, 4, 5]
    while data:
        print(random_pop(data))

# random_pop 함수를 random 모듈의 choice 함수 사용해 만든 것
# random.choice 함수는 입력으로 받은 리스트에서 무작위로 하나 선택해 리턴

1
3
4
2
5


In [14]:
# 리스트 항목 무작위로 섞고 싶을 때, random.sample 함수 사용

import random

data = [1, 2, 3, 4, 5]
random.sample(data, len(data))

# random.sample 두번 째 인수인 len(data)는 무작위로 추출한 원소의 개수를 의미

[1, 3, 2, 5, 4]

In [15]:
import random

data = [1, 2, 3, 4, 5]
random.sample(data, 3)

[5, 3, 1]

* itertools.zip_longest
>  - itertools.zip_longest(*iterables, fillvalue=None) 함수는 같은 개수의 자료형을 묶는 파이썬 내장 함수인 zip 함수와 동일하게 동작
>  - itertools.zip_longest() 함수는 전달한 반복 가능 객체(*iterables)의 길이가 다를 때, 긴 객체의 길이에 맞춰 fillvalue에 설정한 값을 짧은 객체에 채울 수 있음

In [16]:
# itertools_zip.py

students = ['한민서', '황지민', '이영철', '이광수', '김승민']
snacks = ['사탕', '초콜릿', '젤리']

result = zip(students, snacks)
print(list(result))

# students와 snacks의 요소 개수가 다르므로, 더 적은 snacks의 개수만큼만 zip()으로 묶임

[('한민서', '사탕'), ('황지민', '초콜릿'), ('이영철', '젤리')]


In [17]:
# students의 요소 개수가 snacks보다 많을 때, 그만큼을 '새우깡'으로 채우고 싶을 때
# 요소 개수가 많은 것을 기준으로 자료형을 묶는 itertools.zip_longest()를 사용
# 부족한 항목을 None으로 채우기 때문에, fillvalue 값을 '새우깡'으로 지정해 None 대신 다른 값으로 채움

# itertools_zip.py

import itertools

students = ['한민서', '황지민', '이영철', '이광수', '김승민']
snacks = ['사탕', '초콜릿', '젤리']

result = itertools.zip_longest(students, snacks, fillvalue='새우깡')

print(list(result))

[('한민서', '사탕'), ('황지민', '초콜릿'), ('이영철', '젤리'), ('이광수', '새우깡'), ('김승민', '새우깡')]


* itertools.permutation
>  - itertools.permutations(iterable, r)은 반복 가능 객체 중 r개 선택한 순열을 iterator로 리턴
>  - iterator는 반복 가능한 객체


>>  - 1, 2, 3이라는 숫자가 적힌 3장의 카드에서 2장의 카드를 꺼내, 만들 수 있는 2자리 숫자를 모두 구하려면?

>>  - [1, 2, 3]이라는 3장의 카드 중 순서에 상관없이 2장을 뽑는 경우의 수는 모두 3가지 (조합)
>>  - 1, 2
>>  - 2, 3
>>  - 1, 3

>>  - 2자리 숫자이므로, 이 3가지에 순서를 더해 6가지를 만들 수 있음(순열)
>>  - 1,2
>>  - 2,1
>>  - 2,3
>>  - 3,2
>>  - 1,3
>>  - 3,1

In [18]:
import itertools

list(itertools.permutations(['1', '2', '3'], 2))

[('1', '2'), ('1', '3'), ('2', '1'), ('2', '3'), ('3', '1'), ('3', '2')]

In [20]:
for a, b in itertools.permutations(['1', '2', '3'], 2):
    print(a+b)

12
13
21
23
31
32


* 조합을 사용하는 함수
>  - 3장의 카드에서 순서에 상관없이 2장을 고르는 조합은 itertools.combinations() 사용

In [21]:
import itertools

list(itertools.combinations(['1', '2', '3'], 2))

[('1', '2'), ('1', '3'), ('2', '3')]

* itertools.combination
>  - itertools.combinations(iterable, r) 은 반복 가능 객체 중 r개를 선택한 조합을 iterator로 리턴

>>  - 1 ~ 45 중 서로 다른 숫자 6개를 뽑는 로또 번호의 모든 경우의 수(조합)을 구하고, 그 개수를 출력하려면?

In [28]:
import itertools

it = itertools.combinations(range(1, 46), 6)

# 1 ~ 45의 숫자 중에서 6개를 뽑는 경우의 수 iterator로 리턴

In [ ]:
for num in it:
    print(num)

# 실행 시, 끝없이 출력

In [33]:
# iterator 개수만 세려면?

len(list(itertools.combinations(range(1, 46), 6)))

# 로또 번호의 가짓수

8145060

* 중복 조합을 사용하는 함수
>  - 로또 복권이 숫자 중복을 허용하도록 규칙이 변경된다면, 경우의 수는?
>  - 같은 숫자 허용하는 중복 조합은 itertools.combinations_with_replacement()

In [34]:
len(list(itertools.combinations_with_replacement(range(1, 46), 6)))

15890700

* functools.reduce
>  - functools.reduce(function, iterable)은 함수를 반복가능한 객체의 요소에 차례대로 (왼쪽에서 오른쪽) 누적 적용해, 객체를 하나의 값으로 줄이는 함수

In [35]:
# 입력 인수 data의 요소를 모두 더해 리턴하는 add 함수

def add(data):
    result = 0
    for i in data:
        result += i
    return result

data = [1, 2, 3, 4, 5]
result = add(data)
print(result)

15


In [36]:
# functools.reduce() 사용 코드

# reduce_test.py

import functools

data = [1, 2, 3, 4, 5]
result = functools.reduce(lambda x, y: x+y, data)
print(result)

# ((((1+2)+3)+4)+5)
# add 함수와 동일한 역할

15


* functools.reduce()로 최댓값 구하기

In [37]:
num_list = [3, 2, 8, 1, 6, 7]
max_num = functools.reduce(lambda x, y: x if x > y else y, num_list)
print(max_num)

8


* functools.reduce()로 최솟값 구하기

In [38]:
functools.reduce(lambda x, y: x if x < y else y, num_list)

1

* operator.itemgetter
>  - sorted와 같은 함수의 key 매개변수에 적용해, 다양한 기준으로 정렬할 수 있도록 도와주는 모듈

>>  - 학생의 이름, 나이, 성적 등의 정보를 저장한, students 리스트 존재
>>  - students 리스트에는 3개의 튜플이 존재. 각 튜플은 순서대로 이름, 나이, 성적 데이터.
>>  - 이 리스트를 나이순으로 정렬하려면?

>>  - students 가 리스트 일 때

In [ ]:
students = [
    ("jane", 22, 'A'),
    ("dave", 32, 'B'),
    ("sally", 17, 'B')
]

In [39]:
# sorted 함수의 key매개변수에 itemgetter()를 적용하면 해결 가능

# itemgetter1.py

from operator import itemgetter

students = [
    ("jane", 22, 'A'),
    ("dave", 32, 'B'),
    ("sally", 17, 'B')
]

result = sorted(students, key=itemgetter(1))
print(result)

# itemgetter(1)은 students 아이템인 튜플의 두 번째 요소를 기준으로 정렬

[('sally', 17, 'B'), ('jane', 22, 'A'), ('dave', 32, 'B')]


In [1]:
from operator import itemgetter

students = [
    ("jane", 22, 'A'),
    ("dave", 32, 'B'),
    ("sally", 17, 'B')
]

result = sorted(students, key=itemgetter(2))
print(result)

# itemgetter(2)은 students 아이템인 튜플의 세 번째 요소인 성적순으로 정렬

[('jane', 22, 'A'), ('dave', 32, 'B'), ('sally', 17, 'B')]


>>  - students 가 딕셔너리 일 때

In [2]:
# itemgetter2.py

from operator import itemgetter

students = [
    {"name": "jane", "age": 22, "grade": 'A'},
    {"name": "dave", "age": 32, "grade": 'B'},
    {"name": "sally", "age": 17, "grade": 'B' }
]

result = sorted(students, key=itemgetter('age'))
print(result)

# itemgetter('age') 처럼 딕셔너리 키를 사용해야 함

[{'name': 'sally', 'age': 17, 'grade': 'B'}, {'name': 'jane', 'age': 22, 'grade': 'A'}, {'name': 'dave', 'age': 32, 'grade': 'B'}]


In [3]:
from operator import itemgetter

students = [
    {"name": "jane", "age": 22, "grade": 'A'},
    {"name": "dave", "age": 32, "grade": 'B'},
    {"name": "sally", "age": 17, "grade": 'B' }
]

result = sorted(students, key=itemgetter('grade'))
print(result)

[{'name': 'jane', 'age': 22, 'grade': 'A'}, {'name': 'dave', 'age': 32, 'grade': 'B'}, {'name': 'sally', 'age': 17, 'grade': 'B'}]


* operator.attrgetter( )
>  - students 리스트 요소가 튜플이 아닌 Student 클래스의 객체일 때
>  - attrgetter( ) 적용해 정렬해야 함

* 궁금한 점
>  - [<\_\_main\_\_.Student object at 0x7f79ec3983a0>, <\_\_main\_\_.Student object at 0x7f79ec398430>, <\_\_main\_\_.Student object at 0x7f79ec3989a0>] 에서
<\_\_main\_\_.Student object at 0x7f79ec398430> 과 <\_\_main\_\_.Student object at 0x7f79ec44f310> 이 동일하지 않은 이유?
>>  - 리스트 안에 있는 객체와 print(Student('jane', 22, 'A'))에서 생성한 객체는 다르기 때문

>  - age랑 grade로 정렬한 결과값을 보면, 같은 객체가 가지는 주소가 다른 이유?
>>  - 정렬 후 새롭게 result로 할당하기 때문에 주소가 다름

In [13]:
# attrgetter1.py

from operator import attrgetter

class Student:
    def __init__(self, name, age, grade):
        self.name = name
        self.age = age
        self.grade = grade

students = [
    Student('jane', 22, 'A'),
    Student('dave', 32, 'B'),
    Student('sally', 17, 'B')
]

result = sorted(students, key=attrgetter('age'))

print(result)

print(Student('jane', 22, 'A'))

[<__main__.Student object at 0x7f79ec3983a0>, <__main__.Student object at 0x7f79ec398430>, <__main__.Student object at 0x7f79ec3989a0>]


In [14]:
from operator import attrgetter

class Student:
    def __init__(self, name, age, grade):
        self.name = name
        self.age = age
        self.grade = grade

students = [
    Student('jane', 22, 'A'),
    Student('dave', 32, 'B'),
    Student('sally', 17, 'B')
]

result = sorted(students, key=attrgetter('grade'))

print(result)

[<__main__.Student object at 0x7f79ec398d60>, <__main__.Student object at 0x7f79ec398670>, <__main__.Student object at 0x7f79ec398790>]


* shutil
>  - 파일 복사(copy)하거나 이동(move)시 사용하는 모듈

>>  - 작업 중인 파일을 자동으로 백업하는 기능 구현

>>  - /NasData/home/knh/workspace/doit/a.txt 를 /NasData/home/knh/workspace/temp/a.txt.bak 라는 이름으로 복사하는 프로그램
>>  - /NasData/home/knh/workspace/doit 디렉터리에 a.txt 를 만드는 중
>>  - 백업용 /NasData/home/knh/workspace/temp 디렉터리는 이미 생성 완료


In [16]:
# shutil_copy.py

import shutil

shutil.copy("/NasData/home/knh/workspace/doit/a.txt", "/NasData/home/knh/workspace/temp/a.txt.bak")

# temp 디렉터리 아래에 a.txt.bak 생성됨

'/NasData/home/knh/workspace/temp/a.txt.bak'

* shutil.move로 삭제 기능 만들기
>  - /NasData/home/knh/workspace/doit/a.txt 파일을 /NasData/home/knh/workspace/temp/a.txt 로 이동

In [19]:
# shutil_move.py

import shutil

shutil.move("/NasData/home/knh/workspace/doit/a.txt", "/NasData/home/knh/workspace/temp/a.txt")

# doit 안에 있는 a.txt 가 삭제되고, temp 안에 a.txt 생성됨

'/NasData/home/knh/workspace/temp/a.txt'

* glob
>  - 파일을 읽고 쓰는 기능이 있는 프로그램을 만들 시, 특정 디렉터리에 있는 파일 이름을 모두 알아야 할 때가 있음

* 디렉터리에 있는 파일들을 리스트로 만들기 : glob(pathname)
>  - glob 모듈은 디렉터리 안의 파일들을 읽어서 리턴
>  - *, ? 와 같은 메타 문자를 써서 원하는 파일만 읽어 올 수도 있음
>  - ? : 1자리 문자열
>  - \* : 임의의 길이의 문자열

>>  - /NasData/home/knh/workspace 디렉터리에 있는 파일 중 이름이 mark로 시작하는 파일을 모두 찾아 읽어들여보자

In [4]:
import glob

glob.glob("/NasData/home/knh/workspace/mark*")

['/NasData/home/knh/workspace/marks3.py',
 '/NasData/home/knh/workspace/marks1.py',
 '/NasData/home/knh/workspace/marks2.py']

* pickle
>  - pickle은 객체의 형태를 그대로 유지하며, 파일을 저장하고 불러올 수 있게 하는 모듈

>>  - pickle 모듈의 dump 함수를 사용해, 딕셔너리 객체인 data를 그대로 파일에 저장해보기

>> -  궁금한 점
>>>  - the file is not displayed in the text editor because it is either binary of uses an unsupported text encoding ( 파일 내용 확인이 안됨 )
>>>  - vscode가 소스 코드 에디터라서 안됨 ( 밑에 zip을 열어 보는 것도 안되는 것도 이 이유때문 )

In [16]:
import pickle

f = open("test.txt", 'wb')
data = {1: 'python', 2: 'you need'}
pickle.dump(data, f)
f.close()

# text.txt가 생성됨
# the file is not displayed in the text editor because it is either binary of uses an unsupported text encoding

>> - pickle.dump로 저장한 파일을 pickle.load 사용해, 원래 있던 딕셔너리 객체(data) 상태 그대로 불러오기

In [17]:
import pickle

f = open("test.txt", 'rb')
data = pickle.load(f)
print(data)

# 딕셔너리 외 어떤 자료형이든 저장하고 불러올 수 있음

{1: 'python', 2: 'you need'}


In [1]:
import pickle

f = open("test2.txt", 'wb')
data = [1, 'python', 2, 'you need']
pickle.dump(data, f)
f.close()

In [2]:
import pickle

f = open("test2.txt", 'rb')
data = pickle.load(f)
print(data)

[1, 'python', 2, 'you need']


* os
>  - 환경 변수, 디렉터리, 파일 등의 OS 자원을 제어할 수 있게 해주는 모듈

* 내 시스템의 환경 변숫값을 알고 싶을 때 - os.environ
>  - 시스템은 제각기 다른 환경 변숫값을 가지고 있음
>  - os.environ은 현재 시스템의 환경 변숫값을 리턴
>  - os.environ은 환경 변수에 대한 정보를 딕셔너리 형태로 구성된 environ 객체로 리턴
>  - 환경 변수? 프로세스가 컴퓨터에서 동작하는 방식에 영향을 미치는, 동적인 값들의 모임

In [3]:
import os

os.environ

environ{'CONDA_PROMPT_MODIFIER': '(py3.9) ',
        'MAIL': '/var/mail/knh',
        'USER': 'knh',
        'SSH_CLIENT': '155.230.134.92 58168 50305',
        'LD_LIBRARY_PATH': '/usr/local/cuda/extras/CUPTI/lib64',
        'SHLVL': '5',
        'CONDA_SHLVL': '2',
        'HOME': '/NasData/home/knh',
        '_CE_M': '',
        'LOGNAME': 'knh',
        '_': '/NasData/home/knh/anaconda3/envs/py3.9/bin/python',
        '_CE_CONDA': '',
        'PATH': '/NasData/home/knh/anaconda3/envs/py3.9/bin:/NasData/home/knh/.vscode-server/bin/695af097c7bd098fbf017ce3ac85e09bbc5dda06/bin/remote-cli:/NasData/home/knh/anaconda3/envs/py3.9/bin:/NasData/home/knh/bin:/NasData/home/knh/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/usr/local/cuda/bin:/NasData/home/knh/.local/bin:/NasData/home/knh/bin',
        'VSCODE_AGENT_FOLDER': '/NasData/home/knh/.vscode-server',
        'CONDA_PYTHON_EXE': '/NasData/home/knh/anaconda3/bin/python',
 

In [6]:
import os

os.environ['USER']

'knh'

In [7]:
os.environ['PATH']

# PATH 환경 변수 내용

'/NasData/home/knh/anaconda3/envs/py3.9/bin:/NasData/home/knh/.vscode-server/bin/695af097c7bd098fbf017ce3ac85e09bbc5dda06/bin/remote-cli:/NasData/home/knh/anaconda3/envs/py3.9/bin:/NasData/home/knh/bin:/NasData/home/knh/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/usr/local/cuda/bin:/NasData/home/knh/.local/bin:/NasData/home/knh/bin'

* 디렉터리 위치 변경 - os.chdir
>  - 현재 디렉터리의 위치 변경

In [13]:
import os

os.chdir("/NasData/home/knh/mlcl/MLCL_2023")

* 디렉터리 위치 돌려받기 - os.getcwd
>  - 현재 자신의 디렉터리 위치 리턴

In [14]:
os.getcwd()

'/NasData/home/knh/mlcl/MLCL_2023'

In [19]:
import os

os.chdir("/NasData/home/knh/workspace")

In [20]:
os.getcwd()

'/NasData/home/knh/workspace'

* 시스템 명령어 호출 - os.system
>  - 시스템 자체의 프로그램이나 기타 명령어를 파이썬에서 호출 가능
>  - os.system("명령어")

In [6]:
os.system("dir")

0509.py					       cifar10_.py
0604					       colab
0613.ipynb				       data
06132.ipynb				       doit
06133.ipynb				       fig.png
06134.ipynb				       file_with.py
06141.ipynb				       foo.txt
06142.ipynb				       game
06143.ipynb				       image\ processing.py
06144.ipynb				       marks1.py
3-1.ipynb				       marks2.py
3-2.ipynb				       marks3.py
3-3.ipynb				       mod1.py
4-1.ipynb				       modtest.py
4-2.ipynb				       multistring.py
4-3.ipynb				       mymod
4-4.ipynb				       paper.py
5-1.ipynb				       read.py
5-2.ipynb				       read_for.py
5-3.ipynb				       readlines.py
5-4.ipynb				       readlines2.py
5-5.ipynb				       sys1.py
5-6.ipynb				       sys2.py
__pycache__				       temp
add_data.py				       term\ project\ data.py
binary_file.py				       test.txt
cifar\ 10\ resnet\ classifier.py	       test2.txt
cifar10\ image.py			       text
cifar10\ resnet\ classifier\ wuth\ pytorch.py  text.txt
cifar10-baseline			       writedata.py
cifar10.ipynb				     

0

  새파일.txt


* 실행한 시스템 명령어의 결괏값 돌려받기 - os.popen
>  - 시스템 명령어를 실행한 결괏값을 읽기 모드 형태의 파일 객체로 리턴

In [7]:
f = os.popen("dir")

In [8]:
print(f.read())

0509.py					       cifar10_.py
0604					       colab
0613.ipynb				       data
06132.ipynb				       doit
06133.ipynb				       fig.png
06134.ipynb				       file_with.py
06141.ipynb				       foo.txt
06142.ipynb				       game
06143.ipynb				       image\ processing.py
06144.ipynb				       marks1.py
3-1.ipynb				       marks2.py
3-2.ipynb				       marks3.py
3-3.ipynb				       mod1.py
4-1.ipynb				       modtest.py
4-2.ipynb				       multistring.py
4-3.ipynb				       mymod
4-4.ipynb				       paper.py
5-1.ipynb				       read.py
5-2.ipynb				       read_for.py
5-3.ipynb				       readlines.py
5-4.ipynb				       readlines2.py
5-5.ipynb				       sys1.py
5-6.ipynb				       sys2.py
__pycache__				       temp
add_data.py				       term\ project\ data.py
binary_file.py				       test.txt
cifar\ 10\ resnet\ classifier.py	       test2.txt
cifar10\ image.py			       text
cifar10\ resnet\ classifier\ wuth\ pytorch.py  text.txt
cifar10-baseline			       writedata.py
cifar10.ipynb				     

* os 관련 함수
>  - os.mkdir(디렉터리) : 디렉터리 생성
>  - os.rmdir(디렉터리) : 디렉터리 삭제 (디렉터리가 비어 있어야 삭제 가능)
>  - os.remove(파일) : 파일 지움
>  - os.rename(src, dst) : scr라는 이름의 파일을 dst라는 이름으로 바꿈

* zipfile
>   - 여러 개의 파일을 zip 형식으로 합치거나 해제 시 사용하는 모듈

>>  - a.txt, b.txt, c.txt 텍스트 파일을 하나로 합쳐, 'mytext.zip' 이라는 파일을 만들고, 이 파일을 원래의 텍스트 파일 3개로 해제하는 프로그램을 만들려면?

In [30]:
# zipfile_test.py

import zipfile

# 파일 합치기

with zipfile.ZipFile('mytext.zip', 'w') as myzip:
    myzip.write('a.txt')
    myzip.write('b.txt')
    myzip.write('c.txt')

# mytext.zip 안에 a.txt, b.txt, c.txt 생성됨

In [31]:
!unzip mytext.zip -d ./mytext

# 리눅스 코드

Archive:  mytext.zip
 extracting: ./mytext/a.txt          
 extracting: ./mytext/b.txt          
 extracting: ./mytext/c.txt          


In [32]:
# 해제하기

with zipfile.ZipFile('mytext.zip') as myzip:
    myzip.extractall()

>  - ZipFile 객체의 write( ) 함수로 개별 파일 추가 가능
>  - extreactall( ) 함수로 모든 파일 해제 가능

>  - 합친 파일에서 특정 파일만 해제하고 싶을 때, extract( ) 함수 사용

In [33]:
# 특정 파일만 해제하기

with zipfile.ZipFile('mytext.zip') as myzip:
    myzip.extract('a.txt')

>  - 파일을 압축해 묶고 싶은 경우 compression, compresslevel 옵션 사용
>  - with zipfile.ZipFile('mytext.zip', 'w', compression=zipfile.ZIP_LZMA, compresslevel=9) as myzip:

>  - compression 4가지 종류

>  - ZIP_STORED : 압축하지 않고 파일을 zip으로만 묶음. 속도 빠름
>  - ZIP_DEFLATED : 일반적인 zip 압축으로 속도가 빠르고 압축률 낮음 ( 호환성 좋음 )
>  - ZIP_BZIP2 : bzip2 압축으로 압축률이 높고 속도가 느림
>  - ZIP_LZMA : lzma 압축으로 압축률이 높고 속도가 느림 ( 7zip과 동일한 알고리즘 )

>  - compressionlevel
>>  - 압축 수준 의미하는 숫자값으로 1 ~ 9 사용
>>  - 1은 속도가 가장 빠르지만 압축률 낮음
>>  - 9는 속도는 가장 느리지만 압축률 높음

* threading
>  - 스레드 프로그래밍은 초보 프로그래머가 구현하기에는 어려움
>  - 컴퓨터에서 동작하고 있는 프로그램을 프로세스(process)라고 함
>  - 보통 1개의 프로세스는 1가지 일만 함
>  - 스레드(thread)를 사용하면 한 프로세스 안에서 2가지 또는 그 이상의 일 동시 수행 가능

>>  - 스레드의 단점 
>>>  - 같은 변수를 사용할 때, 각 프로세스에서 같은 변수를 동시에 참조하게 됨
>>>  - 변수 값이 계속 바뀌면 프로그램이 꼬일 수도 있는 위험성이 있음


In [35]:
# thread_test.py

import time

def long_task(): # 5초의 시간이 걸리는 함수
    for i in range(5):
        time.sleep(1) # 1초간 대기
        print("working:%s\n" % i)

print("Start")

for i in range(5): # long_task를 5회 수행
    long_task()

print("End")

# long_task는 수행 시 5초의 시간이 걸리는 함수
# 위 프로그램은 이 함수를 총 5번 반복해서 수행하는 프로그램

# 5초가 5번 반복되므로 총 25초의 시간이 걸림

Start
working:0

working:1

working:2

working:3

working:4

working:0

working:1

working:2

working:3

working:4

working:0

working:1

working:2

working:3

working:4

working:0

working:1

working:2

working:3

working:4

working:0

working:1

working:2

working:3

working:4

End


In [37]:
# 스레드 사용 시, 5초의 시간이 걸리는 long_task 함수를 동시에 실행할 수 있어 시간 단축 가능

# thread_test.py

import time
import threading # 스레드 생성하기 위해서는 threading 모듈 필요

def long_task():
    for i in range(5):
        time.sleep(1)
        print("working:%s\n" % i)

print("Start")

threads = []

for i in range(5):
    t = threading.Thread(target = long_task) # 스레드 생성
    threads.append(t)

for t in threads:
    t.start() # 스레드 실행

print("End")

# 25초가 걸리던 작업이 5초 정도에 수행됨
# threading.Thread 사용해 만든 스레드 객체가 동시 작업 가능하게 함

# 프로그램 실행 시, "Start"와 "End"가 먼저 출력되고, 그 이후에 스레드 결과 출력됨
# 프로그램 정상 종료 되지 않음

Start
End


working:0
working:0


working:0

working:0

working:0

working:1

working:1

working:1

working:1

working:1

working:2
working:2


working:2

working:2

working:2

working:3
working:3


working:3

working:3

working:3

working:4
working:4


working:4

working:4

working:4



In [39]:
# "Start"가 출력되고, 그 다음에 스레드의 결과가 출력된 후, 마지막으로 "End"가 출력되게 하려면?
# 스레드의 join 함수를 사용해, 해당 스레드가 종료될 때까지 기다리게 하면 됨

# thread_test.py

import time
import threading

def long_task():
    for i in range(5):
        time.sleep(1)
        print("working:%s\n" % i)

print("Start")

threads = []

for i in range(5):
    t = threading.Thread(target=long_task)
    threads.append(t)

for t in threads:
    t.start()

for t in threads:
    t.join() # join으로 스레드가 종료될 때 까지 기다림

print("End")

Start
working:0

working:0

working:0

working:0

working:0

working:1

working:1

working:1

working:1

working:1

working:2

working:2

working:2

working:2

working:2

working:3
working:3


working:3

working:3

working:3

working:4

working:4

working:4

working:4

working:4

End


* tempfile
>  - 파일을 임시로 만들어 사용할 때 유용한 모듈

>  - tempfile.mkstemp( )
>  - 중복되지 않는 임시 파일의 이름을 무작위로 만들어 리턴

>>  - 궁금한 점
>>  - 이름만 만들어 리턴하는지, 임시로 파일을 만들어 해당 이름으로 리턴하는지?

In [41]:
import tempfile

filename = tempfile.mkstemp()
filename

(67, '/tmp/tmpfbfsoxh2')

>  - tempfile.TemporaryFile( ) 임시 저장 공간으로 사용할 파일 객체를 리턴
>  - 기본적으로 바이너리 쓰기 모드(wb)를 갖음
>  - f.close( ) 호출되면 이 파일은 자동으로 삭제

In [43]:
import tempfile

f = tempfile.TemporaryFile()
f.close()

* traceback
>  - 프로그램 실행 중 발생한 오류를 추적할 때 사용 모듈
>  - 복잡한 파이썬 코드는 어디에서 어떤 오류가 발생했는지 알기 어려움

In [45]:
# traceback_test.py

def a():
    return 1/0

def b():
    a()

def main():
    try:
        b()
    except:
        print("오류가 발생했습니다")

main()

# main()함수 시작하면 b() 함수 호출, b() 함수에서 다시 a() 함수 호출하고, 1을 0으로 나누니까 오류가 발생해서, "오류가 발생했습니다" 메시지 출력

# 코드에서 오류가 발생한 위치와 원인을 정확히 판단해보자

오류가 발생했습니다


In [46]:
# traceback_test.py

import traceback # traceback 모듈 적용

def a():
    return 1/0

def b():
    a()

def main():
    try:
        b()
    except:
        print("오류가 발생했습니다.")
        print(traceback.format_exc()) # traceback 모듈의 format_exc()는 오류 추적 결과를 문자열로 리턴하는 함수

main()

# main()함수에서  b() 함수 호출, b() 함수에서 다시 a() 함수 호출, 1을 0으로 나누니까 ZeroDivisionError 오류가 발생한 것을 로그를 통해 정확하게 확인 가능

오류가 발생했습니다.
Traceback (most recent call last):
  File "/tmp/ipykernel_48679/1929379760.py", line 13, in main
    b()
  File "/tmp/ipykernel_48679/1929379760.py", line 9, in b
    a()
  File "/tmp/ipykernel_48679/1929379760.py", line 6, in a
    return 1/0
ZeroDivisionError: division by zero



* json
>  - JSON 데이터 쉽게 처리하는 모듈

>>  - 개인정보를 JSON 형태의 데이터로 만든 myinfo.json 파일
>>  - 인터넷으로 얻은 이 파일을 읽어 파이썬에서 처리할 수 있도록 딕셔너리 자료형으로 만드려면?
>>  - 즉, JSON 파일을 읽어 딕셔너리로 변환하려면 json 모듈 사용하자

>>>  - 파일명 : myinfo.json

{
    "name": "홍길동",
    "birth": "0525",
    "age": 30
}

>>  - JSON 파일 읽을 때, json.load(파일_객체)를 사용
>>  - load( ) 함수는 읽은 데이터를 딕셔너리 자료형으로 리턴

In [50]:
import json

with open('myinfo.json') as f:
    data = json.load(f)

type(data)

dict

In [51]:
data

{'name': '홍길동', 'birth': '0525', 'age': 30}

>>  - 딕셔너리 자료형을 JSON 파일로 생성 시, json.dump(딕셔너리, 파일_객체) 사용

In [53]:
import json

data = {'name': '홍길동', 'birth': '0525', 'age': 30}

with open('myinfo.json', 'w') as f:
    json.dump(data, f)

# myinfo.json이 생성되고, {"name": "\ud64d\uae38\ub3d9", "birth": "0525", "age": 30} 를 담고 있음

>>  - 딕셔너리 자료형을 JSON 문자열로 만들려면, json.dumps() 함수를 사용

In [57]:
import json

d = {"name": "홍길동", "birth": "0525", "age": 30}
json_data = json.dumps(d)
json_data

'{"name": "\\ud64d\\uae38\\ub3d9", "birth": "0525", "age": 30}'

>>  - dump( ), dumps( ) 함수는 기본적으로 데이터를 아스키 형태로 저장함
>>  - 딕셔너리를 JSON 데이터로 변경하면, '홍길동'과 같은 한글 문자열이 코드 형태로 표시
>>  - 유니코드 문자열을 아스키 형태로 저장하다 보니 한글 문자열이 깨진 것 처럼 보임
>>  - 하지만 JSON 문자열을 딕셔너리로 다시 역변환하는 데는 문제가 없음

>>>  - 유니코드? 컴퓨터 처리를 위한 텍스트 표현에 사용되는 범용 문자 인코딩 표준. 일관성 있는 방식으로 다국어 텍스트를 인코딩할 수 있도록 하며 국제 텍스트 파일 교환을 원활하게 처리함
>>>  - 아스키 형태? 미국에서 정의한 표준화한 부호체계. 7비트 즉, 128개의 고유한 값만 사용

>>  - JSON 문자열을 딕셔너리로 변환할 때, json.loads( ) 함수 사용

In [58]:
json.loads(json_data)

{'name': '홍길동', 'birth': '0525', 'age': 30}

>>  - 한글 문자열이 아스키 형태의 문자열로 변경되는 것을 방지하는 방법
>>  - ensure_ascii = False 옵션 사용
>>  - 데이터 저장 시, 아스키 형태로 변환하지 않겠다는 뜻

In [59]:
d = {"name": "홍길동", "birth": "0525", "age": 30}
json_data = json.dumps(d, ensure_ascii = False)
json_data

'{"name": "홍길동", "birth": "0525", "age": 30}'

In [60]:
json.loads(json_data)

{'name': '홍길동', 'birth': '0525', 'age': 30}

>>  - 출력되는 JSON 문자열 보기 좋게 정렬하려면?
>>  - indent 옵션 추가 ( 들여쓰기 )

In [61]:
d = {"name": "홍길동", "birth": "0525", "age": 30}
print(json.dumps(d, indent=2, ensure_ascii=False))

{
  "name": "홍길동",
  "birth": "0525",
  "age": 30
}


In [64]:
d = {"name": "홍길동", "birth": "0525", "age": 30}
print(json.dumps(d, indent=1, ensure_ascii=False))

{
 "name": "홍길동",
 "birth": "0525",
 "age": 30
}


In [65]:
d = {"name": "홍길동", "birth": "0525", "age": 30}
print(json.dumps(d, indent=3, ensure_ascii=False))

{
   "name": "홍길동",
   "birth": "0525",
   "age": 30
}


>>  - 리스트, 튜플같은 다른 자료형도 JSON 문자열로 바꿀 수 있음

In [66]:
json.dumps([1, 2, 3])

'[1, 2, 3]'

In [67]:
json.dumps((4, 5, 6))

'[4, 5, 6]'

* urllib
>  - URL을 읽고 분석할 때 사용 모듈

>>  - 브라우저로 위키독스의 특정 페이지를 읽고 싶을 때
>>  - https://wikidocs.net/페이지_번호

>>  - 오프라인으로도 읽을 수 있도록, 페이지 번호를 입력받아, 위키독스의 특정 페이지를 wikidocs_페이지_번호.html 파일로 저장하는 함수를 만드는 방법

>>  - URL 호출해 원하는 리소스 얻고 싶을 때, urllib 모듈 사용

In [2]:
# urllib_test.py

import urllib.request

def get_wikidocs(page):
    resource = 'https://wikidocs.net/{}'.format(page)
    with urllib.request.urlopen(resource) as s:
        with open('wikidocs_%s.html' % page, 'wb') as f:
            f.write(s.read())

# get_wikidocs(page)는 위키독스의 페이지 번호를 입력받아, 해당 페이지의 리소스 내용을 파일로 저장하는 함수
# urllib.request.urlopen(resource) 로 s 객체 생성
# s.read()로 리소스 내용 전체를 읽어 저장

In [4]:
get_wikidocs(33)

# https://wikidocs.net/33 웹 페이지를 wikidocs_33.html 이라는 파일로 저장

HTTPError: HTTP Error 403: Forbidden

* 궁금한 점
>  - 왜 오류가 발생했을까?
>  - '403 Forbidden' 는 웹사이트 접속 실패 오류 메시지
>  - 접속하려는 웹 페이지에 대한 액세스가 어떠한 이유로 금지되었음을 나타내는 일반적인 HTTP 상태 이상 코드

* webbrowser
>  - 파이썬 프로그램에서 시스템 브라우저를 호출할 때 사용하는 모듈

>>  - 개발 중 파이썬 문서를 참고하려 함
>>  - https://python.org 사이트를 새로운 웹 브라우저로 열기 위한 코드는?

>>  - 파이썬으로 웹 페이지를 새 창으로 열려면, webbrowser 모듈의 open_new( ) 함수 사용

In [7]:
# webbrowser_test.py

import webbrowser

webbrowser.open_new('http://python.org')

# https://www.python.org/ 창이 뜬다

True

>>  - 이미 열린 브라우저로 원하는 사이트를 열고 싶다면, open_new( ) 대신 open( ) 사용

In [8]:
webbrowser.open('https://wikidocs.net/33')

True